File: algorithmic_toolbox/week2_algorithmic_warmup/6_last_digit_of_the_sum_of_fibonacci_numbers/FibonacciSumLastDigit.java
import java.util.Scanner;

public class FibonacciSumLastDigit {
  private static long getFibonacciSumNaive(long n) {
    if (n <= 1)
      return n;

    long previous = 0;
    long current = 1;
    long sum = 1;

    for (long i = 0; i < n - 1; ++i) {
      long tmp_previous = previous;
      previous = current;
      current = tmp_previous + current;
      sum += current;
    }

    return sum % 10;
  }

  /*
   * The last digit of the numbers in the Fibonacci Sequence form a pattern that
   * repeats after
   * every 60th number. Source:
   * https://www.goldennumber.net/fibonacci-60-repeating-pattern/
   */
  private static long getFibonacciSum(long n) {
    return getFibonacciSumNaive(n % 60);
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    long n = scanner.nextLong();
    scanner.close();

    System.out.println(getFibonacciSum(n));
    System.out.println(getFibonacciSumNaive(n));
  }
}

File: algorithmic_toolbox/week2_algorithmic_warmup/6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.py
import sys


def fibonacci_sum_naive(n: int) -> int:
    if n <= 1:
        return n

    previous, current, _sum = 0, 1, 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        _sum += current

    return _sum % 10


def fibonacci_sum(n: int) -> int:
    """The last digit of the numbers in the Fibonacci Sequence form a pattern that repeats after every 60th number.

    Source: https://www.goldennumber.net/fibonacci-60-repeating-pattern/"""

    return fibonacci_sum_naive(n % 60)


if __name__ == "__main__":
    n = int(sys.stdin.read())
    print(fibonacci_sum(n))
    print(fibonacci_sum_naive(n))
